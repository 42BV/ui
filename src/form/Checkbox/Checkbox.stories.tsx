import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { Alert } from 'reactstrap';

import Checkbox, { JarbCheckbox } from './Checkbox';
import { FinalForm } from '../story-utils';

import { Icon } from '../../core/Icon';
import Tooltip from '../../core/Tooltip/Tooltip';
import { Card } from '../../core/Card/Card';

function isSClass(value?: boolean) {
  if (value !== false) {
    return undefined;
  }

  return 'Hero must be S class to proceed';
}

storiesOf('Form/Checkbox', module)
  .add('basic', () => {
    const [isSClass, setIsSClass] = useState<boolean | undefined>(undefined);

    return (
      <Card className="m-2">
        <Checkbox
          id="isSClass"
          label="Is S class hero"
          placeholder="Whether or not the hero is S class"
          value={isSClass}
          onChange={setIsSClass}
        />

        {isSClass ? (
          <Alert color="success" className="d-flex">
            <Icon icon="warning" className="mr-1" /> This hero is of the highest
            caliber
          </Alert>
        ) : null}
      </Card>
    );
  })
  .add('basic with indeterminate', () => {
    const [isSClass, setIsSClass] = useState<boolean | undefined>(undefined);

    return (
      <Card className="m-2">
        <Checkbox
          id="isSClass"
          label="Is S class hero"
          placeholder="Whether or not the hero is S class"
          value={isSClass}
          onChange={setIsSClass}
          allowIndeterminate={true}
        />

        {isSClass ? (
          <Alert color="success" className="d-flex">
            <Icon icon="warning" className="mr-1" /> This hero is of the highest
            caliber
          </Alert>
        ) : null}
      </Card>
    );
  })
  .add('without placeholder', () => {
    const [isSClass, setIsSClass] = useState<boolean | undefined>(undefined);

    return (
      <Card className="m-2">
        <Checkbox
          id="isSClass"
          label="Is S class hero"
          value={isSClass}
          onChange={setIsSClass}
        />

        {isSClass ? (
          <Alert color="success" className="d-flex">
            <Icon icon="warning" className="mr-1" /> This hero is of the highest
            caliber
          </Alert>
        ) : null}
      </Card>
    );
  })
  .add('with custom label', () => {
    const [isSClass, setIsSClass] = useState<boolean | undefined>(undefined);

    return (
      <Card className="m-2">
        <Checkbox
          id="isSClass"
          label={
            <div className="d-flex justify-content-between">
              <span>Is S class hero</span>
              <Tooltip
                className="ml-1"
                content="An S class hero is a hero of the highest caliber"
              >
                <Icon icon="info" />
              </Tooltip>
            </div>
          }
          placeholder="Whether or not the hero is S class"
          value={isSClass}
          onChange={setIsSClass}
        />

        {isSClass ? (
          <Alert color="success" className="d-flex">
            <Icon icon="warning" className="mr-1" /> This hero is of the highest
            caliber
          </Alert>
        ) : null}
      </Card>
    );
  })
  .add('jarb', () => {
    return (
      <FinalForm>
        <JarbCheckbox
          id="isSClass"
          name="isSClass"
          label="Is S class hero"
          placeholder="Whether or not the hero is S class"
          validators={[isSClass]}
          jarb={{
            validator: 'Hero.isSClass',
            label: 'Description'
          }}
        />
      </FinalForm>
    );
  })
  .add('jarb with indeterminate', () => {
    return (
      <FinalForm>
        <JarbCheckbox
          id="isSClass"
          name="isSClass"
          label="Is S class hero"
          placeholder="Whether or not the hero is S class"
          validators={[isSClass]}
          jarb={{
            validator: 'Hero.isSClass',
            label: 'Description'
          }}
          allowIndeterminate={true}
        />
      </FinalForm>
    );
  });
