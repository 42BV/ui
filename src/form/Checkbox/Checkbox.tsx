import React from 'react';
import { FormGroup, Input, Label } from 'reactstrap';

import withJarb from '../withJarb/withJarb';
import { Color } from '../..';
import { doBlur } from '../utils';

type Props = {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The value that the form element currently has.
   *
   * When the value is `true` the checkbox is checked.
   * When the value `false` the checkbox is unchecked.
   *
   * When the value is `undefined` it depends on `allowIndeterminate`,
   * when `allowIndeterminate` is `true` the checkbox is indeterminate.
   * when  `allowIndeterminate` is `false` the checkbox is unchecked.
   */
  value?: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional callback for when the form element is blurred.
   */
  onBlur?: () => void;

  /**
   * Optionally the error message to render.
   */
  error?: React.ReactNode;

  /**
   * Optionally the color of the FormGroup.
   */
  color?: Color;

  /**
   * Whether or not the form element is currently valid.
   */
  valid?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * The label of the form element.
   */
  label: React.ReactNode;

  /**
   * Optionally the placeholder of the form element.
   */
  placeholder?: string;

  /**
   * Optionally whether to support the indeterminate state.
   */
  allowIndeterminate?: boolean;
};

/**
 * Checkbox is a form element for when the value is a boolean.
 */
export default function Checkbox(props: Props) {
  const {
    id,
    onChange,
    onBlur,
    error,
    color,
    label,
    className = '',
    placeholder,
    value,
    valid,
    allowIndeterminate
  } = props;

  const checked = !!value;

  const invalid = valid === false ? true : undefined;

  function onClick() {
    onChange(!checked);
    doBlur(onBlur);
  }

  return (
    <FormGroup check className={className} color={color}>
      <Label for={id} check>
        <Input
          id={id}
          type="checkbox"
          checked={checked}
          onChange={onClick}
          valid={valid}
          invalid={invalid}
          innerRef={(e) => {
            if (e && allowIndeterminate) {
              // Value will be an empty string when coming from
              // final-form. So we do a ts-expect-error to go around
              // the fact that we say value is boolean | undefined.

              // @ts-expect-error Accept that a checkbox event has an indeterminate
              e.indeterminate = value === undefined || value === '';
            }
          }}
        />{' '}
        {label}
      </Label>

      {placeholder ? (
        <p className="text-muted mb-0">
          <em>{placeholder}</em>
        </p>
      ) : null}

      {error}
    </FormGroup>
  );
}

/**
 * Variant of the Checkbox which can be used in a Jarb context.
 */
export const JarbCheckbox = withJarb<boolean, boolean | null, Props>(Checkbox);
