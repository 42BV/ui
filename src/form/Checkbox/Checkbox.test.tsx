import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { Checkbox } from './Checkbox';

describe('Component: Checkbox', () => {
  function setup({
    value,
    hasPlaceholder,
    allowIndeterminate,
    hasLabel
  }: {
    value?: boolean;
    hasPlaceholder?: boolean;
    allowIndeterminate?: boolean;
    hasLabel?: boolean;
  }) {
    const onChangeSpy = jest.fn();
    const onBlurSpy = jest.fn();

    const props = {
      id: hasLabel ? 'agree' : undefined,
      label: 'Agree',
      hiddenLabel: !hasLabel,
      placeholder: hasPlaceholder ? 'Do you agree' : undefined,
      value,
      valid: value === true,
      allowIndeterminate,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      error: 'Some error',
    };

    const { container, rerender, asFragment } = render(
      <Checkbox {...props} />
    );

    return { container, rerender, asFragment, onBlurSpy, onChangeSpy };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('with placeholder', () => {
      setup({ hasPlaceholder: true });
      expect(screen.queryByText('Do you agree')).toBeInTheDocument();
    });

    test('without placeholder', () => {
      setup({ hasPlaceholder: false });
      expect(screen.queryByText('Do you agree')).not.toBeInTheDocument();
    });

    test('visible label', () => {
      setup({ hasLabel: true });
      expect(screen.queryByText('Agree')).toBeInTheDocument();
      expect(screen.queryByLabelText('Agree')).toBeInTheDocument();
    });

    test('invisible label', () => {
      setup({ hasLabel: false });
      expect(screen.queryByText('Agree')).not.toBeInTheDocument();
      expect(screen.queryByLabelText('Agree')).toBeInTheDocument();
    });
  });

  describe('allowIndeterminate is true', () => {
    it('should set indeterminate to false when value is true', () => {
      setup({ value: true, allowIndeterminate: true });
      expect(screen.getByRole('checkbox')).not.toBePartiallyChecked();
      expect(screen.getByRole('checkbox')).toBeChecked();
    });

    it('should set indeterminate to false when value is false', () => {
      setup({ value: false, allowIndeterminate: true });
      expect(screen.getByRole('checkbox')).not.toBePartiallyChecked();
      expect(screen.getByRole('checkbox')).not.toBeChecked();
    });

    it('should set indeterminate to true when value is undefined', () => {
      setup({ value: undefined, allowIndeterminate: true });
      expect(screen.getByRole('checkbox')).toBePartiallyChecked();
      expect(screen.getByRole('checkbox')).not.toBeChecked();
    });
  });

  describe('events', () => {
    test('onChange', () => {
      const { onBlurSpy, onChangeSpy } = setup({ value: true });

      fireEvent.click(screen.getByLabelText('Agree'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(false);

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('value changes', () => {
    test('becomes empty', () => {
      const { rerender, onBlurSpy, onChangeSpy } = setup({ value: true });

      expect(screen.getByRole('checkbox')).toBeChecked();

      rerender(
        <Checkbox
          id="agree"
          label="Agree"
          value={undefined}
          valid={true}
          onChange={onChangeSpy}
          onBlur={onBlurSpy}
          error="Some error"
        />
      );

      expect(screen.getByRole('checkbox')).not.toBeChecked();
    });

    test('becomes filled from false', () => {
      const { rerender, onChangeSpy, onBlurSpy } = setup({ value: false });

      expect(screen.getByRole('checkbox')).not.toBeChecked();

      rerender(
        <Checkbox
          id="agree"
          label="Agree"
          value={true}
          valid={true}
          onChange={onChangeSpy}
          onBlur={onBlurSpy}
          error="Some error"
        />
      );

      expect(screen.getByRole('checkbox')).toBeChecked();
    });

    test('becomes filled from undefined', () => {
      const { rerender, onBlurSpy, onChangeSpy } = setup({ value: undefined });

      expect(screen.getByRole('checkbox')).not.toBeChecked();

      rerender(
        <Checkbox
          id="agree"
          label="Agree"
          value={true}
          valid={true}
          onChange={onChangeSpy}
          onBlur={onBlurSpy}
          error="Some error"
        />
      );

      expect(screen.getByRole('checkbox')).toBeChecked();
    });
  });
});
