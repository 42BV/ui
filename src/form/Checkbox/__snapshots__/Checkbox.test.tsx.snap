// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Checkbox ui with placeholder: Component: Checkbox => ui => with placeholder 1`] = `
<FormGroup
  check={true}
  className=""
  tag="div"
>
  <Label
    check={true}
    for="agree"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
        "xxl",
      ]
    }
  >
    <Input
      checked={true}
      id="agree"
      innerRef={[Function]}
      onChange={[Function]}
      type="checkbox"
      valid={true}
    />
     
    Agree
  </Label>
  <p
    className="mb-0"
  >
    <em>
      Do you agree
    </em>
  </p>
  Some error
</FormGroup>
`;

exports[`Component: Checkbox ui without placeholder: Component: Checkbox => ui => without placeholder 1`] = `
<FormGroup
  check={true}
  className=""
  tag="div"
>
  <Label
    check={true}
    for="agree"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
        "xxl",
      ]
    }
  >
    <Input
      checked={false}
      id="agree"
      innerRef={[Function]}
      invalid={true}
      onChange={[Function]}
      type="checkbox"
      valid={false}
    />
     
    Agree
  </Label>
  Some error
</FormGroup>
`;
